storageconnector/gitstorage.go

```diff
@@ -0,0 +1,270 @@
+package storageconnector
+
+import (
+	"strings"
+	"path"
+	"sort"
+	"gopkg.in/yaml.v2"
+	"gopkg.in/src-d/go-git.v4"
+	"gopkg.in/src-d/go-git.v4/core"
+	"io/ioutil"
+	"fmt"
+)
+
+const (
+	featureDir = "features"		// name of the directory where features are located.
+	featureFile = "meta.yml"	// name of the file containing all metadata for a feature.
+	featureSnippet = "Dockerfile"   // the file containing the actual docker snippet.
+	defaultSearchParamsLimit = 100
+)
+
+// yamlFeatureMeta is used for unmarshalling of meta.yml files.
+type yamlFeatureMeta struct {
+	Description string
+	Dependencies []string
+}
+
+// gitStorage is an implementation of StorageReader based on
+// a git repository as storage back-end.
+type gitStorage struct {
+	repo 	*git.Repository
+}
+
+// NewStorageReader returns a StorageReader which uses a public git repository
+// as data source for pazuzu features.
+//
+// url:  The URL to the git repository that serves as data source. The
+//       repository must be publicly accessible.
+//
+// If the repository can't be accessed NewStorageReader returns an error.
+func NewStorageReader(url string) (StorageReader, error) {
+	// OPTIMIZATION: can be an fs repository which is cached and only pulled when needed
+	repo := git.NewMemoryRepository()
+
+	err  := repo.Clone(&git.CloneOptions{
+		URL: url,
+		ReferenceName: core.HEAD,
+		SingleBranch: true,
+		Depth: 1,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &gitStorage{repo: repo}, nil
+}
+
+func (storage *gitStorage) SearchMeta(params SearchParams) ([]FeatureMeta, error) {
+	commit, err := storage.latestCommit()
+	if err != nil {
+		return nil, err
+	}
+
+	all, err := commit.Files()
+	if err != nil {
+		return nil, err
+	}
+
+	// find matching feature names
+	matchedNames := map[string]bool{}
+	matchedNamesList := []string{}
+	all.ForEach(func(file *git.File) error {
+		pathComponents := strings.Split(file.Name, "/")
+
+		// check if file is in feature dir
+		if pathComponents[0] != featureDir {
+			return nil
+		}
+
+		// check if feature was already found
+		featureName := pathComponents[1]
+		if matchedNames[featureName] {
+			return nil
+		}
+
+		// check if feature matches search params
+		if params.Name.MatchString(featureName) {
+			matchedNames[featureName] = true
+			matchedNamesList = append(matchedNamesList, featureName)
+		}
+
+		return nil
+	})
+
+	if params.Limit == 0 {
```

Comment:
Is it correct that first all the matching features are put into `matchedNamesList` and only after that `Limit` is applied? In such case the code might process over 100500 matches (assume they do exist) and then take first 100 (defined by `Limit`). Looks like you need to move Limit processing into `all.ForEach` construct or refactor it to take `Limit` value in account.